### PR TITLE
HDDS-8530. [Snapshot] Fix for OM crash on restart due to snapshot chain manager corruption

### PR DESCRIPTION
**Pull request by hemantk-12**
## What changes were proposed in this pull request?
OM was crashing on the restart because snapshot chain was getting corrupted. On the deep dive, it is found that it is because creation time of snapshots are not unique among snapshots and causing the failure. Mode details are in jira: https://issues.apache.org/jira/browse/HDDS-8530.
This change is to pass the creation time through [preExecute](https://github.com/apache/ozone/blob/master/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/snapshot/OMSnapshotCreateRequest.java#L92) instead of each OM node using current time as creation time. Creation time is passed similar to `snapshotId`.

This change also includes some logging changes in `SnapshotChainManager` to log more information to debug in future.

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-8530

## How was this patch tested?
Added integration test to verify.

### CODE REVIEW

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOzoneManagerHASnapshot.java [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+ */
+@Timeout(300)
+public class TestOzoneManagerHASnapshot extends TestOzoneManagerHA {
+
+  /**
+   * Test snapshotNames are unique among OM nodes when snapshotName is not
+   * passed or empty.
+   */
+  @Test
+  public void testUniqueSnapshotName() throws Exception {
```
Comment (smengcl) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOzoneManagerHASnapshot.java:51]: Same applies here https://github.com/apache/ozone/pull/4657/files/480b998aca193b001912674fe1504d4c44792fd3#r1195799002

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOzoneManagerHASnapshot.java [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+    }
+
+    // Restart leader OM
+    OzoneManager omLeader = getCluster().getOMLeader();
+    getCluster().shutdownOzoneManager(omLeader);
+    getCluster().restartOzoneManager(omLeader, true);
+
+    await().atMost(Duration.ofSeconds(180))
+        .until(() -> getCluster().getOMLeader() != null);
+  }
```
Comment (smengcl) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOzoneManagerHASnapshot.java:119]: hmm. Is there supposed to be snapshot chain restore/reload validation logic towards the end of this test case?

**Review by smengcl — COMMENTED** [review]
Thanks @hemantk-12 for the fix!

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOzoneManagerHASnapshot.java (hunk 1 above)
Comment (hemantk-12) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOzoneManagerHASnapshot.java]: Thanks. Changed the test name.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOzoneManagerHASnapshot.java (hunk 2 above)
Comment (hemantk-12) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOzoneManagerHASnapshot.java]: Added assertion that OMLeader node should be non-null.

OM restart will fail if snapshotChainRestore fails as mentioned in https://issues.apache.org/jira/browse/HDDS-8530?focusedCommentId=17720063&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17720063

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/SnapshotChainManager.java [hunk 3]
```diff
@@ -499,23 +468,24 @@ public String nextPathSnapshot(String snapshotPath, String snapshotID)
...
         .get(snapshotPath)
         .get(snapshotID)
         .getPreviousSnapshotID() == null) {
-      LOG.error("no preceeding snapshot for provided snapshotPath {}, "
-          + "snapshotID {}", snapshotPath, snapshotID);
-      throw new NoSuchElementException("no preceeding snapshot from: "
-          + snapshotID + "for snapshot path:" + snapshotPath);
+      LOG.error("No preceding snapshot for provided snapshotId: {} and " +
+          "snapshotPath: {}", snapshotPath, snapshotID);
+      throw new NoSuchElementException(String.format("Ne preceding snapshot " +
```
Comment (aswinshakil) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/SnapshotChainManager.java:486]: nit:
```suggestion
      throw new NoSuchElementException(String.format("No preceding snapshot " +
```

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/SnapshotChainManager.java [hunk 4]
```diff
@@ -499,23 +468,24 @@ public String nextPathSnapshot(String snapshotPath, String snapshotID)
...
     if (snapshotChainPath
         .get(snapshotPath)
         .get(snapshotID)
         .getPreviousSnapshotID() == null) {
-      LOG.error("no preceeding snapshot for provided snapshotPath {}, "
-          + "snapshotID {}", snapshotPath, snapshotID);
-      throw new NoSuchElementException("no preceeding snapshot from: "
-          + snapshotID + "for snapshot path:" + snapshotPath);
+      LOG.error("No preceding snapshot for provided snapshotId: {} and " +
+          "snapshotPath: {}", snapshotPath, snapshotID);
```
Comment (aswinshakil) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/SnapshotChainManager.java:485]:
```suggestion
          "snapshotPath: {}", snapshotID, snapshotPath);
```

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/SnapshotChainManager.java [hunk 5]
```diff
@@ -468,26 +439,24 @@ public boolean hasPreviousPathSnapshot(String snapshotPath, String snapshotID)
...
-        .get(snapshotID)
+    if (snapshotChainPath.get(snapshotPath).get(snapshotID)
         .getNextSnapshotID() == null) {
-      LOG.error("no following snapshot for provided snapshotPath {}, "
-          + "snapshotID {}", snapshotPath, snapshotID);
-      throw new NoSuchElementException("no following snapshot from: "
-          + snapshotID + "for snapshot path:" + snapshotPath);
+      LOG.error("No following snapshot for provided snapshotId {} and " +
+          "snapshotPath {}.", snapshotID, snapshotPath);
+      throw new NoSuchElementException(String.format("Ne following snapshot " +
```
Comment (aswinshakil) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/SnapshotChainManager.java:455]:
```suggestion
      throw new NoSuchElementException(String.format("No following snapshot " +
```

**Review by aswinshakil — COMMENTED** [review]
Thanks for working this @hemantk-12 . LGTM. Just have a few minor comments.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/SnapshotChainManager.java (hunk 3 above)
Comment (hemantk-12) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/SnapshotChainManager.java]: Thanks for detailed review @aswinshakil. Updated as suggested.

**Comment by smengcl** [comment]
Thanks @aswinshakil for reviewing this.

**Pull request merged**